website/src/_posts/2019-05-01-adding-data-sources-to-sql.md

```diff
@@ -0,0 +1,174 @@
+---
+layout: post
+title:  "Adding new Data Sources to Beam SQL CLI"
+date:   2019-05-01 00:00:01 -0800
+excerpt_separator: <!--more-->
+categories: blog
+authors:
+        - pabloem
+
+---
+<!--
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+A new, exciting feature that came to Apache Beam is the ability to use
+SQL in your pipelines. This is done using Beam's
+[`SqlTransform`](https://beam.apache.org/releases/javadoc/latest/org/apache/beam/sdk/extensions/sql/SqlTransform.html)
+in Java pipelines.
+
+Beam also has a fancy new SQL command line that you can use to query your
+data interactively, be it Batch or Streaming. If you haven't tried it, check out
+[http://bit.ly/ExploreBeamSQL](http://bit.ly/ExploreBeamSQL).
+
+A nice feature of the SQL CLI is that you can use `CREATE TABLE` commands to
+*add* data sources to be accessed in the CLI. Currently, the CLI supports
+creating tables from BigQuery, PubSub, Kafka, and text files. In this post, we
+explore how to add new data sources, so that you will be able to consume data
+from other Beam sources.
+
+<!--more-->
+
+Beam's `SqlTransform` works by relying on `TableProvider`s, which it uses when
```

Comment:
I would put a full `CREATE TABLE` example here so that the users will understand the goal before diving into implementation, e.g. something along these lines:

The table provider we will be implementing in this post will be generating a continuous unbounded stream of integers. It will be based on GenerateSequence PTransform (link) from the Beam SDK. In the end will be able to define and use the sequence generator in SQL like this:

 * define the corresponding external table:

```sql
CREATE EXTERNAL TABLE .                    -- all tables in Beam are external, they are not persisted
  sequenceTable                              -- table alias that will be used in queries
  (
         sequence LONG,                    -- sequence number
         event_timestamp TIMESTAMP         -- timestamp of the generated event
   )
TYPE sequence                              -- type identifies the table provider
TBLPROPERTIES '{ elementsPerSecond : 12 }' -- optional rate at which events are generated
```
* use it in queries:
```sql
SELECT sequence FROM sequenceTable;
```

Reply:
I've added this. LMK what you think.